ST_resgroups.c

```diff
@@ -185,124 +181,124 @@ void rgroup_PartResources(void)
-			if (add_seeds)
-			{
-				rgroup_AddSpecies(rg, sp);
-				Species_Update_Newsize(sp, newsize);
-			}
-		}
-
-		if (add_seeds && !forced)
-			_add_annual_seedprod(sp, (ZRO(x)) ? -1. : g_pr);
-
-		sumsize += newsize;
-	}
-
-	/*if (g->regen_ok == TRUE) {
-	 printf("Regen_ok %u\n",rg);}*/
-
-	/*will print out a flag of Regen_ok with the resourse group number (numbered
-	 * in the order of turned on groups in the "rgroup.in" file */
-
-	return (sumsize / (RealF) g->max_spp);
-
+static RealF _add_annuals(const GrpIndex rg, const SppIndex sp, const RealF lastyear_relsize) {
+    /*======================================================*/
+    /* check regen_ok flag.  if true, apply establishment and
+     * add to seedbank.  Otherwise, add 0 to seedbank and skip
+     * adding plants this year.  
+     * reset group size to account for additions, or 0 if none.
+     */
+	#define xF_DELTA (20*F_DELTA)
+	#define xD_DELTA (20*D_DELTA)
+	#define ZERO(x) \
+    ( (sizeof(x) == sizeof(float)) \
+  ? ((x)>-xF_DELTA && (x)<xF_DELTA) \
+  : ((x)>-xD_DELTA && (x)<xD_DELTA) )
+    IntU i, num_est; //number of individuals that will establish this year
+    RealF viable_seeds, viable_seeds_reduced; //viable seed and viable seed excluding this year's added seed, respectively
+    float alpha, beta, var; //parameters needed to calculation of num_est from beta distribution
+    GroupType *g;
+    SpeciesType *s;
+
+    g = RGroup[rg];
+    assert(g->max_age == 1);
+
+    s = Species[sp];
+    viable_seeds_reduced = 0.0;
+
+    /*Increment viable years for seeds and add seeds to the seedbank*/
+    _add_annual_seedprod(sp, lastyear_relsize);
+
+    /*Implement decay process and then get viable seeds from seed bank*/
+    viable_seeds = (g->regen_ok) ? _get_annual_maxestab(sp) : 0;
+
+    /*Create a beta random number draw, where mean =s->seedling_estab_prob
+     variance = s->var. 
+     Calculate alpha and beta based on the mean and variance values by species.in*/
+     alpha = (pow (s->seedling_estab_prob, 2) - pow (s->seedling_estab_prob, 3)
+             - s->seedling_estab_prob * s->var) / s->var;
+     beta = (s->seedling_estab_prob - pow (s->seedling_estab_prob, 3) 
+             - s->var + s->var * s->seedling_estab_prob)/ s->var;
+     
+     /*Create beta random number*/
+     var = RandBeta(alpha, beta);
+     
+     /*Determine number of seedlings to add*/
+     num_est = viable_seeds * var;
+     	//printf("Species name=%s , num_est   =%u \n",s->name,  num_est);
+     
+     /*if the beta random number * viable seeds is larger than max_seed_estab, 
+     max_seed_estab value is used instead*/
+     if (num_est > s->max_seed_estab) {num_est = s->max_seed_estab;} 
+          //printf("Species name=%s , num_est   =%u \n",s->name,  num_est);
+    
+    /*Remove seedlings added from the seed bank*/
+    /*Calculate the viable seedbank excluding seeds added this year*/
+    viable_seeds_reduced= (IntU) viable_seeds - s->seedprod[0];
```

Comment:
This line 237 and lines 243-248 mean that seeds produced from the previous year productivity do not germinate in the current year? Despite that `num_est` is based on a value that includes those seeds? I believe that most species' seeds germinate in the spring following the season when they were produced.

Reply:
Yes, seeds that are added this year s->seedprod[0] are currently protected from removal. I understand the point you are making here and tend to agree. Are you proposing to do the following instead:

/*Multiple the proportion of seeds in each viable year array by the total 
    number of seeds that germinated as seedlings and subtract those seeds from 
    the relevant seedprod array.*/
       for (i = 0; i<= s->viable_yrs; i++)
       {
        s->seedprod[i] =  s->seedprod[i] -  round(num_est * s->seedprod[i] / viable_seeds); 
       }

Reply:
Addressed by https://github.com/DrylandEcology/STEPWAT2/commit/1bcb6ce8c3278a122e27905c720cdefcd247803a